src/vdom/component-recycler.js

```diff
@@ -33,13 +33,13 @@ export function createComponent(Ctor, props, context) {
 		inst;
 
 	if (Ctor.prototype && Ctor.prototype.render) {
-		inst = new Ctor(props, context);
+		inst = new /** @type {(props, context) => void} */(Ctor)(props, context);
```

Comment:
This seems strange to me. @rbiggs Can you elaborate on the thinking behind it?

Reply:
That's a type cast. We're casting it to a basic function to avoid type error displayed in this image:

<img width="698" alt="ctor" src="https://user-images.githubusercontent.com/249218/42736387-e39e2990-881a-11e8-8aa0-ad45802e8fbc.png">

This is a common problem when dealing with constructor functions and TypeScript. A simple type cast resolve this problem.

Reply:
The following change I proposed in rbiggs/preact#2 should make this read a bit better:

```diff
--- a/src/vdom/component-recycler.js
+++ b/src/vdom/component-recycler.js
@@ -33,7 +33,9 @@ export function createComponent(Ctor, props, context) {
                inst;
 
        if (Ctor.prototype && Ctor.prototype.render) {
-               inst = new /** @type {(props, context) => void} */(Ctor)(props, context);
+               const ComponentCtor =
+                       /** @type {(props, context) => void} */(Ctor);
+               inst = new ComponentCtor (props, context);
                Component.call(inst, props, context);
        }
        else {
```

Reply:
The variables are fine, it's the typings that I have an issue with. It seems wrong and more like an issue with typescript if one has to define a class constructor as a plain function.

Reply:
Actually, it's the wrong type being defined in the parameter:


```javascript
/**
 * Create a component. Normalizes differences between PFC's and classful
 * Components.
 * @param {function} Ctor The constructor of the component to create
 * @param {object} props The initial props of the component
 * @param {object} context The initial context of the component
 * @returns {import('../component').Component}
 */
```

The Ctor type is described as a generic function, which can't be used with the `new` keyword. Just need to change that to the same thing we used for the type cast:

```javascript
/**
 * Create a component. Normalizes differences between PFC's and classful
 * Components.
 * @param {(props, context) => void} Ctor The constructor of the component to create
 * @param {object} props The initial props of the component
 * @param {object} context The initial context of the component
 * @returns {import('../component').Component}
 */
```

Then the type cast isn't necessary:

```javascript
if (Ctor.prototype && Ctor.prototype.render) {
  inst = new Ctor(props, context);
  Component.call(inst, props, context);
}
```